src/extension/geofence/src/main/java/org/geoserver/geofence/GeofenceAccessManager.java

```diff
@@ -269,108 +279,189 @@ public StyleAccessLimits getAccessLimits(Authentication user, StyleInfo style) {
+        else
+            accessLimits =
+                    buildResourceAccessLimits(
+                            ((LayerInfo) info).getResource(), AccessInfo.ALLOW_ALL, null);
+        return accessLimits;
+    }
 
-        if (request != null) {
-            if ("*".equals(request)) {
-                ruleFilter.setRequest(RuleFilter.SpecialFilterType.ANY);
-            } else {
-                ruleFilter.setRequest(request);
-            }
+    private Pair<Geometry, AccessInfo> getAllowedAreaAndAccessInfoFromContainers(
+            CatalogInfo resource, List<LayerGroupInfo> containers, Authentication user) {
+        boolean directAccess = containers == null || containers.isEmpty();
+        Request req = Dispatcher.REQUEST.get();
+        String service = req != null ? req.getService() : null;
+        boolean isWms = service != null && service.equalsIgnoreCase("WMS");
+        Pair<Geometry, AccessInfo> containerRule = null;
+        if (directAccess && isWms) {
+            containerRule = getContainerRuleForLayerDirectAccess(resource, user);
         } else {
-            ruleFilter.setRequest(RuleFilter.SpecialFilterType.DEFAULT);
+            containerRule = handleContainers(containers, user);
         }
-        ruleFilter.setWorkspace(workspace);
-        ruleFilter.setLayer(layer);
+        return containerRule;
+    }
 
-        String sourceAddress = retrieveCallerIpAddress();
-        if (sourceAddress != null) {
-            ruleFilter.setSourceAddress(sourceAddress);
-        } else {
-            LOGGER.log(Level.WARNING, "No source IP address found");
-            ruleFilter.setSourceAddress(RuleFilter.SpecialFilterType.DEFAULT);
+    private String getUserNameFromAuth(Authentication authentication) {
+        String username = authentication != null ? authentication.getName() : null;
+        if (username != null && username.isEmpty()) {
+            username = null;
         }
+        return username;
+    }
 
-        LOGGER.log(Level.FINE, "ResourceInfo filter: {0}", ruleFilter);
-
-        AccessInfo rule = rules.getAccessInfo(ruleFilter);
-
-        if (rule == null) {
-            rule = AccessInfo.DENY_ALL;
+    private Pair<Geometry, AccessInfo> handleContainers(
+            List<LayerGroupInfo> groups, Authentication user) {
+        Geometry allowedArea = null;
+        AccessInfo lessRestrictiveRule = null;
+        for (LayerGroupInfo lgi : groups) {
+            WorkspaceInfo ws = lgi.getWorkspace();
+            String workspace = ws != null ? ws.getName() : null;
+            String layer = lgi.getName();
+            RuleFilter filter = buildRuleFilter(workspace, layer, user);
+            AccessInfo accessInfo = rules.getAccessInfo(filter);
+            lessRestrictiveRule = getLessRestrictiveRule(lessRestrictiveRule, accessInfo);
+            allowedArea =
+                    getReprojectArea(accessInfo, lgi.getBounds().getCoordinateReferenceSystem());
```

Comment:
So for the rule, the less restrictive is taken. But the area is just the last one found instead? What if two groups have different restrictions? Are there tests for that?

Reply:
fixed